### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.loggers.comet`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to `pytorch_lightning.loggers.comet` for issue #13445

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/loggers/csv_logs.py [hunk 1]
```diff
@@ -195,7 +195,7 @@ def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
         self.experiment.log_hparams(params)
 
     @rank_zero_only
-    def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
+    def log_metrics(self, metrics: Dict[str, Union[Tensor, float]], step: Optional[int] = None) -> None:
```
Comment (JustinGoheen) [inline src/pytorch_lightning/loggers/csv_logs.py:198]: updated to use `metrics: Dict[str, Union[Tensor, float]]` across all loggers for `metrics`.

File: src/pytorch_lightning/utilities/logger.py [hunk 2]
```diff
@@ -132,7 +132,9 @@ def _sanitize_params(params: Dict[str, Any]) -> Dict[str, Any]:
     return params
 
 
-def _add_prefix(metrics: Dict[str, float], prefix: str, separator: str) -> Dict[str, float]:
+def _add_prefix(
```
Comment (JustinGoheen) [inline src/pytorch_lightning/utilities/logger.py:135]: relevant to using `metrics: Dict[str, Union[Tensor, float]]` across all loggers for `metrics`.

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/loggers/comet.py [hunk 3]
```diff
@@ -326,7 +328,7 @@ def version(self) -> str:
 
         return self._future_experiment_key
 
-    def __getstate__(self):
+    def __getstate__(self):  # type: ignore[no-untyped-def]
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/comet.py:331]:
```suggestion
    def __getstate__(self) -> Dict[str, Any]

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/utilities/logger.py (hunk 2 above)
Comment (otaj) [reply src/pytorch_lightning/utilities/logger.py:135]: Hi @JustinGoheen, because of this change, this PR now doesn't pass `mypy` check

Comment (JustinGoheen) [reply src/pytorch_lightning/utilities/logger.py:135]: updated; apologies for not fixing this earlier.

File: src/pytorch_lightning/loggers/comet.py [hunk 4]
```diff
@@ -141,14 +141,16 @@ def __init__(
         prefix: str = "",
         agg_key_funcs: Optional[Mapping[str, Callable[[Sequence[float]], float]]] = None,
         agg_default_func: Optional[Callable[[Sequence[float]], float]] = None,
-        **kwargs,
+        **kwargs: Any,
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/comet.py:144]: @mauvilsa We are setting the type for the kwargs here. However, then the test in `test_cli.py::test_comet_logger_init_args` fails with
```

E               ValueError: Value "Namespace(class_path='pytorch_lightning.loggers.CometLogger', init_args=Namespace(save_dir='comet', workspace='comet'))" does not validate against any of the types in typing.Union[pytorch_lightning.loggers.logger.Logger, typing.Iterable[pytorch_lightning.loggers.logger.Logger], bool]:
E                 - expected AST, got 'NoneType'
E                 - Expected a List but got "Namespace(class_path='pytorch_lightning.loggers.CometLogger', init_args=Namespace(save_dir='comet', workspace='comet'))"
E                 - Expected a <class 'bool'> but got "Namespace(class_path='pytorch_lightning.loggers.CometLogger', init_args=Namespace(save_dir='comet', workspace='comet'))"
```

https://github.com/Lightning-AI/lightning/runs/7476365010?check_suite_focus=true

Do you know what the correct annotation is here? Perhaps we should drop it and ignore the mypy error?

Comment (mauvilsa) [reply src/pytorch_lightning/loggers/comet.py:144]: The issue is not the type of `**kwargs` but a bug in jsonargparse and the

Comment (mauvilsa) [reply src/pytorch_lightning/loggers/comet.py:144]: The issue is not the type of `**kwargs` but a bug in jsonargparse and the `self._save_dir: Optional[str]`, `self.rest_api_key: Optional[str]` lines below. This bug was fixed in https://github.com/omni-us/jsonargparse/pull/148 and has already been released. But with the new version a test will fail. I created a pull request to fix the test https://github.com/Lightning-AI/lightning/pull/13805.

Comment (awaelchli) [reply src/pytorch_lightning/loggers/comet.py:144]: Thank you!

**Pull request merged**